### PR TITLE
(Re)process runs from the GUI

### LINKED ISSUE #39: Allow the users to reprocess runs from the GUI
Reprocessing runs is one of the few times we force users to use a terminal, which is scary for lots of them. It should be possible to reprocess runs and specific variables from the GUI.

### PR DESCRIPTION
**Pull request by takluyver**
Another attempt at this, now that processing happens in Slurm.

image

Closes #39

### CODE REVIEW

**Comment by tmichela** [comment]
Could you also add an option on right click?

**Comment by tmichela** [comment]
it'd also be nice to check if new variables exist in the context file that were never processed, otherwise it will always be skipped, unless one selects all variables.

File: damnit/gui/process.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        main_vbox.addLayout(hbox1)
+
+        grid1 = QtWidgets.QGridLayout()
+        hbox1.addLayout(grid1)
+        vbox2 = QtWidgets.QVBoxLayout()
+        hbox1.addLayout(vbox2)
+
+        self.edit_prop = QtWidgets.QLineEdit(proposal)
+        #self.edit_prop.setInputMask('999900;_')  # 4-6 digits
+        grid1.addWidget(QtWidgets.QLabel("Proposal:"), 0, 0)
```
[the remark this answers is not in the document]
Comment (takluyver) [reply damnit/gui/process.py:84]: I'm trying to preserve the possibility of one DAMNIT table covering multiple proposals, without getting in the way of the common case - the proposal number should be pre-filled if possible, and the run field gets the focus. Do you prefer to push people back to the CLI if multiple proposals ever comes up.

**Comment by takluyver** [comment]
I'll add a right-click option. :+1: 

For new variables: would it be a decent compromise to update the variables whenever you successfully save the context file in the GUI? Loading the context file often takes a few seconds, so doing it when showing this dialog either means an awkward pause before showing it, or a janky update after the list shows up, neither of which seem like great UX.

I'm also wondering about an option to process only variables which don't already have results, to fill in newly added variables.

File: damnit/gui/table.py [hunk 2]
```diff
@@ -60,6 +60,8 @@ def __init__(self) -> None:
         self.show_logs_action = QtWidgets.QAction('View processing logs')
         self.show_logs_action.triggered.connect(self.show_run_logs)
         self.context_menu.addAction(self.show_logs_action)
+        self.process_action = QtWidget.QAction('Reprocess runs')
```
Comment (tmichela) [inline damnit/gui/table.py:63]:
```suggestion
        self.process_action = QtWidgets.QAction('Reprocess runs')
```

**Comment by tmichela** [comment]
> For new variables: would it be a decent compromise to update the variables whenever you successfully save the context file in the GUI? 

It is really that long to validate the file? 🤔  
What do you think of having a "load new variables" button then on the widget? or a "process new variable" checkbox?

File: damnit/gui/process.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmichela) [reply damnit/gui/process.py:84]: I honestly hope we won't have to support this case 😅 but sure, why not...

**Comment by tmichela** [comment]
I'm thinking that I would be nice to have a visual indicator of the runs processing, particularly in this case of processing from the gui. But I think we don't have a way to monitor slurm jobs status yet, right?

**Comment by takluyver** [comment]
> It is really that long to validate the file? 🤔

It's now common/normal to use a different Python environment for evaluating the context file from running the DAMNIT application. So validating it means launching a subprocess and then loading all the modules it imports. Between GPFS, which isn't great for lots of small files, and xarray, which gets its `__version__` in a convenient but inefficient way, it can easily be a couple of seconds. Not a long time, but definitely not instant for GUI code, hence #291.

> I'm thinking that I would be nice to have a visual indicator of the runs processing

Yup, I'm also planning that, but I wanted to do this first.

File: damnit/gui/process.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            self.vars_list.addItem(itm)
+
+        self.vars_list.itemChanged.connect(self.validate_vars)
+
+        self.validate_runs()
+        self.validate_vars()
+
+        self.edit_runs.setFocus()
+
+    def validate_runs(self):
```
Comment (tmichela) [inline damnit/gui/process.py:127]: I would probably be useful to cehck if the selected runs exist, and:
- show a warning message if they don't exist
- do not start a job for runs that do not exist
- show an error if trying to reprocess and none of the selected runs exist

Comment (takluyver) [reply damnit/gui/process.py:127]: OK, how about like this...

1. All chosen runs valid:

image

2. Some don't exist:

image

3. No runs exist, or the proposal number is wrong, or the user running the GUI doesn't have access:

image

In the last case, the 'Ok' button is disabled so you have to change it to launch reprocessing.

Comment (tmichela) [reply damnit/gui/process.py:140]: Great!

**Comment by tmichela** [comment]
I'm happy with the current state, feel free to merge.
We can address the other points later if they are needed.

LGTM

**Comment by takluyver** [comment]
Thanks for the review!

**Pull request merged**